### PR TITLE
fix(agw): Converted s6a task files to cpp

### PR DESCRIPTION
**Pull request by VinashakAnkitAman**
## Title
fix(agw): Converted s6a task files to cpp [Extension to Github Issue: #13096]

## Summary
Migrated all the files related to s6a task in the folders `magma/lte/gateway/c/core/oai/tasks/s6a/` and `magma/lte/gateway/c/core/oai/include/` to `cpp/hpp` and fixed all the compilation errors and warnings.

**Note:**  The function `fd_msg_browse()` is a macro function expanding to `fd_msg_browse_internal()` in the file `/usr/local/include/freeDiameter/libfdproto.h`, which passes the 3rd argument of function by casting to (void *), but received as (msg_or_avp **). This is failing with c++ explicit typecasting validation. To fix this all the `fd_msg_browse()` calls have been shifted to `fd_msg_browse_internal()` function calls to avoid intermediate type conversion by the macro and the 3rd argument is now passed as typecasted value to (msg_or_avp **)

For reference: https://github.com/Metaswitch/freeDiameter/blob/master/include/freeDiameter/libfdproto.h#:~:text=int%20fd_msg_browse_internal%20(%20msg_or_avp,*)(found)%2C%20(depth)%20)

Also followed the guidelines for order of headers from https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

## Test plan
Verified compilation with following:
make clean; make build_oai;
make clean; make test_oai;
make clean; make test_oai BUILD_TYPE=Debug OAI_TESTS=test_amf_app
make clean; bazel build --config=production //lte/gateway/c/core:agw_of
make clean; bazel test //lte/gateway/c/core/oai/test/amf:amf_stateless_test
make clean; make FEATURES=mme_oai build_oai
make clean; export FEATURES=mme_oai; make build_oai

Verified with sanity and in CI

Note: After using `FEATURES=mme_oai;` flag, make test_oai started failing in spgw module with undefined reference to some config functions. Needed to exit out from the VM and re-login to fix the issue.

Signed-off-by: VinashakAnkitAman <ankit.aman@radisys.com>

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/s6a/s6a_auth_info.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_auth_info.cpp]: Move this header out of `extern C` block

**Review by ssanadhya — COMMENTED** [review]
Minor comments. 

Please add in summary why you need to use `fd_msg_browse_internal` and make sure that Magma-OAI-Jenkins CI check is passing with this change.

File: lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp]: Replace this with #pragma once
This comment is applicable for all headers changed in this PR

File: lte/gateway/c/core/oai/tasks/s6a/s6a_iface.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_iface.cpp]: This can be moved outside of extern C

File: lte/gateway/c/core/oai/tasks/s6a/s6a_messages.hpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_messages.hpp]: Include the header files as mentioned in https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: lte/gateway/c/core/oai/tasks/s6a/s6a_peer.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_peer.cpp]: Remove dynamic_memory_check.h file from extern C

File: lte/gateway/c/core/oai/tasks/s6a/s6a_peer.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_peer.cpp:44]: Leave space after extern C

File: lte/gateway/c/core/oai/tasks/s6a/s6a_subscription_data.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_subscription_data.cpp:34]: Put these files also under extern C.

File: lte/gateway/c/core/oai/tasks/s6a/s6a_subscription_data.cpp
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/s6a/s6a_subscription_data.cpp]: Address cpp lint errors

File: lte/gateway/c/core/oai/tasks/s6a/s6a_supported_features.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_supported_features.cpp:32]: Leave space

File: lte/gateway/c/core/oai/tasks/s6a/s6a_task.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_task.cpp]: Include it under extern C

File: lte/gateway/c/core/oai/tasks/s6a/s6a_up_loc.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_up_loc.cpp]: Need not include under extern C

File: lte/gateway/c/core/oai/tasks/s6a/s6a_up_loc.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_up_loc.cpp]: Need not include under extern C

File: lte/gateway/c/core/oai/tasks/s6a/s6a_viface.hpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s6a/s6a_viface.hpp:20]: Remove from extern C

File: lte/gateway/c/core/oai/test/amf/util_s6a_update_location.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/amf/util_s6a_update_location.cpp:14]: Leave space

File: lte/gateway/c/core/oai/test/s6a_task/s6a_recv_async_grpc_messages.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/s6a_task/s6a_recv_async_grpc_messages.cpp:20]: Remove from extern C

File: lte/gateway/c/core/oai/lib/s6a_proxy/s6a_client_api.hpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/lib/s6a_proxy/s6a_client_api.hpp:33]: All these fun(), s6a_purge_ue, s6a_authentication_info_req and s6a_update_location_req need not be under extern C

File: lte/gateway/c/core/oai/test/amf/util_s6a_update_location.hpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/amf/util_s6a_update_location.hpp:16]: Leave space

File: lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp
Comment (VinashakAnkitAman) [reply lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp]: I found some statement on wiki which is not very supportive for use of #pragma once:
https://en.wikipedia.org/wiki/Pragma_once#:~:text=Caveats%5Bedit,pragma%20once.

@ssanadhya What is your opinion on this?

Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp]: @VinashakAnkitAman , the caveats highlighted on the Wiki (such as symbolic links, networked file systems, etc.) do not apply to Magma codebase. So let's continue using the `#pragma once` directive, since it makes it easier for contributors.

**Review by ssanadhya — COMMENTED** [review]
Lgtm, can be merged once @rsarwad approves.

File: lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp
Comment (VinashakAnkitAman) [reply lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.hpp]: Sure @ssanadhya.. Thanks for the clarification. I will update the PR.

File: lte/gateway/c/core/oai/include/messages_def.h
Comment (rsarwad) [inline lte/gateway/c/core/oai/include/messages_def.h:45]: Can we add #pragma once to this file as well

File: lte/gateway/c/core/oai/test/amf/util_s6a_update_location.hpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/amf/util_s6a_update_location.hpp:17]: Include #pragma once to this file

**Review by LKreutzer — APPROVED** [review]
Bazel changes lgtm

File: lte/gateway/c/core/oai/include/messages_def.h
Comment (VinashakAnkitAman) [reply lte/gateway/c/core/oai/include/messages_def.h:45]: This file is included more than once in the file intertask_interface_types.h and it is included not just as a header file, but included within enum and structure. Hence, it is not possible to have "#pragma once" in this file

I have added the same comment in this file as well.

**Review by rsarwad — APPROVED** [review]
LGTM

**Pull request merged**
